contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,817 @@
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+pragma solidity 0.4.24;
+pragma experimental "v0.5.0";
+
+import { ReentrancyGuard } from "zeppelin-solidity/contracts/ReentrancyGuard.sol";
+import { Math } from "zeppelin-solidity/contracts/math/Math.sol";
+import { SafeMath } from "zeppelin-solidity/contracts/math/SafeMath.sol";
+import { Margin } from "../../Margin.sol";
+import { MathHelpers } from "../../../lib/MathHelpers.sol";
+import { TokenInteract } from "../../../lib/TokenInteract.sol";
+import { MarginCommon } from "../../impl/MarginCommon.sol";
+import { LoanOfferingVerifier } from "../../interfaces/LoanOfferingVerifier.sol";
+import { OnlyMargin } from "../../interfaces/OnlyMargin.sol";
+import { CancelMarginCallDelegator } from "../../interfaces/lender/CancelMarginCallDelegator.sol";
+/* solium-disable-next-line max-len*/
+import { ForceRecoverCollateralDelegator } from "../../interfaces/lender/ForceRecoverCollateralDelegator.sol";
+import { IncreaseLoanDelegator } from "../../interfaces/lender/IncreaseLoanDelegator.sol";
+import { LoanOwner } from "../../interfaces/lender/LoanOwner.sol";
+import { MarginCallDelegator } from "../../interfaces/lender/MarginCallDelegator.sol";
+import { MarginHelper } from "../lib/MarginHelper.sol";
+
+
+/**
+ * @title BucketLender
+ * @author dYdX
+ *
+ * On-chain shared lender that allows anyone to deposit tokens into this contract to be used to
+ * lend tokens for a particular position.
+
+ * - Deposits go into a particular bucket, determined by time since the start of the position.
+ * - When lending money, earlier buckets are used to lend first.
+ * - When money is paid back, later buckets are paid back first.
+ * - Over time, this gives higher interest to earlier buckets, but locks-up those funds for longer.
+ * - Deposits in the same bucket earn the same interest.
+ * - Lenders can withdraw their funds at any time if the funds are not being lent.
+ */
+contract BucketLender is
+    OnlyMargin,
+    LoanOwner,
+    IncreaseLoanDelegator,
+    MarginCallDelegator,
+    CancelMarginCallDelegator,
+    ForceRecoverCollateralDelegator,
+    LoanOfferingVerifier,
+    ReentrancyGuard
+{
+    using SafeMath for uint256;
+
+    // ============ Events ============
+
+    // TODO
+
+    // ============ State Variables ============
+
+    // Available token to lend
+    mapping(uint256 => uint256) public availableForBkt;
+    uint256 public availableTotal;
+
+    // Current allocated principal for each bucket
+    mapping(uint256 => uint256) public principalForBkt;
```

Comment:
Just say bucket not bkt, it's 3 more letters on a long thing, and this makes it more confusing

Reply:
We don't use the pattern of abbreviating names anywhere else either